cluster_client.py

```diff
@@ -0,0 +1,98 @@
+import asyncio
+import random
+from dataclasses import dataclass, field
+from typing import Self
+
+import grpc
+from grpc.aio._call import AioRpcError
+
+from rpc.raft_pb2_grpc import ClusterStub
+from rpc.raft_pb2 import GetValueResponse, Key, Mapping, ProposeMappingResponse
+
+
+@dataclass(slots=True)
+class RpcClient:
+    port: int = field()
+    channel: grpc.aio.Channel = field(init=False)
+    stub: ClusterStub = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.channel = grpc.aio.insecure_channel(f"localhost:{self.port}")
+        self.stub = ClusterStub(self.channel)
+
+    async def close(self: Self) -> None:
+        await self.channel.close()
+
+    async def get_value_rpc(self: Self, key: str) -> GetValueResponse:
+        return await self.stub.GetValue(Key(key=key))
+
+    async def propose_mapping_rpc(
+        self: Self, key: str, value: str
+    ) -> ProposeMappingResponse:
+        return await self.stub.ProposeMapping(Mapping(key=key, value=value))
+
+
+@dataclass(slots=True)
+class Client:
+    _all_raft_server_ports: list[int] = field()
+    _current_leader: int | None = field(default=None)
+    _rpc_clients_by_port: dict[int, RpcClient] = field(default_factory=dict)
+
+    def __post_init__(self: Self) -> None:
+        self._rpc_clients_by_port = {
+            # Make sure to use the cluster port for the RpcClient
+            port: RpcClient(port + len(self._all_raft_server_ports))
+            for port in self._all_raft_server_ports
+        }
+        self._choose_a_random_server()
+        self._print(f"Initial random leader is {self._current_leader}")
+
+    def _choose_a_random_server(self: Self) -> None:
+        self._current_leader = self._all_raft_server_ports[
+            random.randrange(len(self._all_raft_server_ports))
+        ]
+        self._print(f"New randomly chosen leader is {self._current_leader}")
+
+    def _print(self: Self, msg: str) -> None:
+        print(f"Client: {msg}")
+
+    def current_leader(self: Self) -> int | None:
```

Comment:
you could make this a property with the @property decorator and then use it elsewhere

is the intention just to make sure users of the class don't modify it?